File: go/test/stop_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/transportrest-transit-apis-sdk"
	"github.com/voxgig-sdk/transportrest-transit-apis-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestStopEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Stop(nil)
		if ent == nil {
			t.Fatal("expected non-nil StopEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := stopBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "stop." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRANSPORTRESTTRANSITAPIS_TEST_STOP_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		stopRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.stop", setup.data)))
		var stopRef01Data map[string]any
		if len(stopRef01DataRaw) > 0 {
			stopRef01Data = core.ToMapAny(stopRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = stopRef01Data

		// LOAD
		stopRef01Ent := client.Stop(nil)
		stopRef01MatchDt0 := map[string]any{
			"id": stopRef01Data["id"],
		}
		stopRef01DataDt0Loaded, err := stopRef01Ent.Load(stopRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		stopRef01DataDt0LoadResult := core.ToMapAny(stopRef01DataDt0Loaded)
		if stopRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if stopRef01DataDt0LoadResult["id"] != stopRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func stopBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "stop", "StopTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read stop test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse stop test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"stop01", "stop02", "stop03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRANSPORTRESTTRANSITAPIS_TEST_STOP_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRANSPORTRESTTRANSITAPIS_TEST_STOP_ENTID": idmap,
		"TRANSPORTRESTTRANSITAPIS_TEST_LIVE":      "FALSE",
		"TRANSPORTRESTTRANSITAPIS_TEST_EXPLAIN":   "FALSE",
		"TRANSPORTRESTTRANSITAPIS_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TRANSPORTRESTTRANSITAPIS_TEST_STOP_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TRANSPORTRESTTRANSITAPIS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TRANSPORTRESTTRANSITAPIS_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTransportrestTransitApisSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRANSPORTRESTTRANSITAPIS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRANSPORTRESTTRANSITAPIS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/StopEntityTest.php
<?php
declare(strict_types=1);

// Stop entity test

require_once __DIR__ . '/../transportresttransitapis_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class StopEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TransportrestTransitApisSDK::test(null, null);
        $ent = $testsdk->Stop(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = stop_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "stop." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TRANSPORTREST_TRANSIT_APIS_TEST_STOP_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $stop_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.stop")));
        $stop_ref01_data = null;
        if (count($stop_ref01_data_raw) > 0) {
            $stop_ref01_data = Helpers::to_map($stop_ref01_data_raw[0][1]);
        }

        // LOAD
        $stop_ref01_ent = $client->Stop(null);
        $stop_ref01_match_dt0 = [
            "id" => $stop_ref01_data["id"],
        ];
        $stop_ref01_data_dt0_loaded = $stop_ref01_ent->load($stop_ref01_match_dt0, null);
        $stop_ref01_data_dt0_load_result = Helpers::to_map(is_object($stop_ref01_data_dt0_loaded) && method_exists($stop_ref01_data_dt0_loaded, 'data_get') ? $stop_ref01_data_dt0_loaded->data_get() : $stop_ref01_data_dt0_loaded);
        $this->assertNotNull($stop_ref01_data_dt0_load_result);
        $this->assertEquals($stop_ref01_data_dt0_load_result["id"], $stop_ref01_data["id"]);

    }
}

function stop_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/stop/StopTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TransportrestTransitApisSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["stop01", "stop02", "stop03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TRANSPORTREST_TRANSIT_APIS_TEST_STOP_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TRANSPORTREST_TRANSIT_APIS_TEST_STOP_ENTID" => $idmap,
        "TRANSPORTREST_TRANSIT_APIS_TEST_LIVE" => "FALSE",
        "TRANSPORTREST_TRANSIT_APIS_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TRANSPORTREST_TRANSIT_APIS_TEST_STOP_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TRANSPORTREST_TRANSIT_APIS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new TransportrestTransitApisSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TRANSPORTREST_TRANSIT_APIS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TRANSPORTREST_TRANSIT_APIS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
